File: test/aidbox_sdk/snapshots/python/hl7_fhir_r4_core/SubstanceReferenceInformation.py
from typing import Optional, List
from dataclasses import dataclass, field
from base import CodeableConcept
from base import Range
from base import Quantity
from base import DomainResource
from base import Reference
from base import Identifier
from base import BackboneElement

@dataclass(kw_only=True)
class SubstanceReferenceInformationGene(BackboneElement):
    gene: Optional[CodeableConcept] = None
    source: Optional[List[Reference]] = field(default_factory=list)
    gene_sequence_origin: Optional[CodeableConcept] = None

@dataclass(kw_only=True)
class SubstanceReferenceInformationTarget(BackboneElement):
    organism: Optional[CodeableConcept] = None
    organism_type: Optional[CodeableConcept] = None
    amount_type: Optional[CodeableConcept] = None
    type: Optional[CodeableConcept] = None
    interaction: Optional[CodeableConcept] = None
    source: Optional[List[Reference]] = field(default_factory=list)
    amount_quantity: Optional[Quantity] = None
    amount_string: Optional[str] = None
    target: Optional[Identifier] = None
    amount_range: Optional[Range] = None

@dataclass(kw_only=True)
class SubstanceReferenceInformationGeneElement(BackboneElement):
    type: Optional[CodeableConcept] = None
    source: Optional[List[Reference]] = field(default_factory=list)
    element: Optional[Identifier] = None

@dataclass(kw_only=True)
class SubstanceReferenceInformationClassification(BackboneElement):
    domain: Optional[CodeableConcept] = None
    source: Optional[List[Reference]] = field(default_factory=list)
    subtype: Optional[List[CodeableConcept]] = field(default_factory=list)
    classification: Optional[CodeableConcept] = None

@dataclass(kw_only=True)
class SubstanceReferenceInformation(DomainResource):
    gene: Optional[List[SubstanceReferenceInformationGene]] = field(default_factory=list)
    target: Optional[List[SubstanceReferenceInformationTarget]] = field(default_factory=list)
    comment: Optional[str] = None
    gene_element: Optional[List[SubstanceReferenceInformationGeneElement]] = field(default_factory=list)
    classification: Optional[List[SubstanceReferenceInformationClassification]] = field(default_factory=list)

File: test/aidbox_sdk/snapshots/dotnet/hl7-fhir-r4-core/SubstanceReferenceInformation.cs
using Aidbox.FHIR.Base;
using Aidbox.FHIR.Utils;

namespace Aidbox.FHIR.R4.Core;

public class SubstanceReferenceInformation : DomainResource, IResource
{
    public SubstanceReferenceInformation_Gene[]? Gene { get; set; }
    public SubstanceReferenceInformation_Target[]? Target { get; set; }
    public string? Comment { get; set; }
    public SubstanceReferenceInformation_GeneElement[]? GeneElement { get; set; }
    public SubstanceReferenceInformation_Classification[]? Classification { get; set; }

    public class SubstanceReferenceInformation_Gene : BackboneElement
    {
        public Base.CodeableConcept? Gene { get; set; }
        public Base.ResourceReference[]? Source { get; set; }
        public Base.CodeableConcept? GeneSequenceOrigin { get; set; }
    }

    public class SubstanceReferenceInformation_Target : BackboneElement
    {
        public object? Amount    
        {
            get
            {
                if (AmountQuantity is not null)
                {
                    return AmountQuantity;
                }
        
                if (AmountString is not null)
                {
                    return AmountString;
                }
        
                if (AmountRange is not null)
                {
                    return AmountRange;
                }
        
                return null;
            }
        
            set
            {
                if (value?.GetType() == typeof(Base.Quantity))
                {
                    AmountQuantity = (Base.Quantity)value;return;
                }
        
                if (value?.GetType() == typeof(string))
                {
                    AmountString = (string)value;return;
                }
        
                if (value?.GetType() == typeof(Base.Range))
                {
                    AmountRange = (Base.Range)value;return;
                }
        
                throw new ArgumentException("Invalid type provided");
            }
        }
        public Base.CodeableConcept? Organism { get; set; }
        public Base.CodeableConcept? OrganismType { get; set; }
        public Base.CodeableConcept? AmountType { get; set; }
        public Base.CodeableConcept? Type { get; set; }
        public Base.CodeableConcept? Interaction { get; set; }
        public Base.ResourceReference[]? Source { get; set; }
        public Base.Quantity? AmountQuantity { get; set; }
        public string? AmountString { get; set; }
        public Base.Identifier? Target { get; set; }
        public Base.Range? AmountRange { get; set; }
    }

    public class SubstanceReferenceInformation_GeneElement : BackboneElement
    {
        public Base.CodeableConcept? Type { get; set; }
        public Base.ResourceReference[]? Source { get; set; }
        public Base.Identifier? Element { get; set; }
    }

    public class SubstanceReferenceInformation_Classification : BackboneElement
    {
        public Base.CodeableConcept? Domain { get; set; }
        public Base.ResourceReference[]? Source { get; set; }
        public Base.CodeableConcept[]? Subtype { get; set; }
        public Base.CodeableConcept? Classification { get; set; }
    }
}